pycyphal/transport/udp/_ip/_endpoint_mapping.py

```diff
@@ -11,98 +11,109 @@
+    return ty(msb | node_id)  # type: ignore
 
 
-def unicast_ip_to_node_id(local_ip_address: IPAddress, node_ip_address: IPAddress) -> typing.Optional[int]:
+def multicast_group_to_service_data_specifier(
+    local_ip_address: IPAddress, multicast_group: IPAddress
+) -> typing.Optional[ServiceDataSpecifier]:
     """
-    Returns the node-ID if the node IP address and the local IP address belong to the same subnet.
-    Returns None if the node is not a member of the local subnet.
-    Raises a value error if either address is a multicast group address.
+    The inverse of :func:`service_data_specifier_to_multicast_group`.
+    The local IP address is needed to ensure that the multicast group belongs to the correct Cyphal/UDP subnet.
+    The return value is None if the multicast group is not valid per the current Cyphal/UDP specification
+    or if it belongs to a different Cyphal/UDP subnet.
 
     >>> from ipaddress import ip_address
-    >>> unicast_ip_to_node_id(ip_address('127.42.1.1'), ip_address('127.42.1.200'))
-    456
-    >>> unicast_ip_to_node_id(ip_address('127.0.0.99'), ip_address('127.0.0.99'))
-    99
-    >>> unicast_ip_to_node_id(ip_address('127.99.1.1'), ip_address('127.42.1.200'))  # Returns None
-    >>> unicast_ip_to_node_id(ip_address('239.42.1.1'), ip_address('127.42.1.200'))
-    Traceback (most recent call last):
-      ...
-    ValueError: Multicast group address cannot be a local IP address...
-    >>> unicast_ip_to_node_id(ip_address('127.42.1.1'), ip_address('239.42.1.200'))
-    Traceback (most recent call last):
-      ...
-    ValueError: Multicast group address cannot be mapped to a node-ID...
+    >>> multicast_group_to_service_data_specifier(ip_address('127.168.11.22'), ip_address('239.41.1.200'))
+    ServiceDataSpecifier(subject_id=456, role=...REQUEST...)
+    >>> multicast_group_to_service_data_specifier(ip_address('127.42.11.22'), ip_address('239.43.1.200'))    # -> None
+    >>> multicast_group_to_service_data_specifier(ip_address('127.42.11.22'), ip_address('239.42.255.200'))  # -> None
     """
-    if local_ip_address.is_multicast:
-        raise ValueError(f"Multicast group address cannot be a local IP address: {local_ip_address}")
-    if node_ip_address.is_multicast:
-        raise ValueError(f"Multicast group address cannot be mapped to a node-ID: {node_ip_address}")
-    if (int(local_ip_address) | IP_ADDRESS_NODE_ID_MASK) == (int(node_ip_address) | IP_ADDRESS_NODE_ID_MASK):
-        return int(node_ip_address) & IP_ADDRESS_NODE_ID_MASK
+    try:
+        # QUESTION: Add a third option UNDEFINED to role?
+        candidate = ServiceDataSpecifier(int(multicast_group) & NODE_ID_MASK, ServiceDataSpecifier.Role.REQUEST)
+    except ValueError:
+        return None
+    # QUESTION: correct that node_id is equal to service_id?
+    if service_data_specifier_to_multicast_group(local_ip_address, candidate.service_id) == multicast_group:
+        return candidate
     return None
```

Comment:
@pavel-kirienko 

Could you comment on these 2 questions.

Reply:
1. This function needs more context to construct a service data specifier. The missing part is the destination UDP port number which can be deconstructed into the service-ID and the role (request or response) using `udp_port_to_service_data_specifier`. Alternatively, and perhaps better, the function should be modified to only return the destination node-ID (it would need to be renamed, though), leaving the rest to the caller.

2. Node-ID and service-ID are completely orthogonal! This is explained in section 4.1.1

Reply:
```py
def multicast_group_to_node_id(
    local_ip_address: IPAddress, multicast_group: IPAddress
) -> typing.Optional[int]:
    """
    The inverse of :func:`service_data_specifier_to_multicast_group`.
    The local IP address is needed to ensure that the multicast group belongs to the correct Cyphal/UDP subnet.
    The return value is None if the multicast group is not valid per the current Cyphal/UDP specification
    or if it belongs to a different Cyphal/UDP subnet.

    >>> from ipaddress import ip_address
    >>> multicast_group_to_node_id(ip_address('127.168.11.22'), ip_address('239.41.1.200'))
    456
    >>> multicast_group_to_node_id(ip_address('127.42.11.22'), ip_address('239.43.1.200'))    # -> None
    >>> multicast_group_to_node_id(ip_address('127.42.11.22'), ip_address('239.42.255.200'))  # -> None
    """
    candidate = int(multicast_group) & NODE_ID_MASK
    if service_data_specifier_to_multicast_group(local_ip_address, candidate) == multicast_group:
        return candidate
    return None
```

Like this?

Reply:
Oh but wait, there's one problem: the local IP address should no longer be relied on at all because in the new design it is irrelevant for communication. Instead, we should pass the Cyphal/UDP domain-ID (aka subnet ID) as an explicit integer in [0,32). A new parameter will need to be added to the top-level API (the UDPTransport constructor) with a default value of zero.

Reply:
Oh but wait, there's one problem: the local IP address should no longer be relied on at all because in the new design it is irrelevant for communication. Instead, we should pass the Cyphal/UDP domain-ID (aka subnet ID) as an explicit integer in [0,32). A new parameter will need to be added to the top-level API (the UDPTransport constructor) with a default value of zero.

Reply:
Ok, so to make sure I understand correctly:

With this port to multicast, there's no need for the property `local_ip_address` which can be replaced by `domain_id`. 

`_endpoint_mapping.py` will contain the following functions (related to `service`):

```py
service_data_specifier_to_multicast_group(domain_id: int, node_id: int) -> IPAddress
service_multicast_group_to_node_id(domain_id: int, multicast_group: IPAddress) -> typing.Optional[int]

service_data_specifier_to_udp_port(ds: ServiceDataSpecifier) -> int
udp_port_to_service_data_specifier(port: int) -> typing.Optional[ServiceDataSpecifier]
```

And the same can be said for the functions related to `message`?

```py
message_data_specifier_to_multicast_group(domain_id: int, data_specifier: MessageDataSpecifier) -> IPAddress
multicast_group_to_message_data_specifier(domain_id: int, multicast_group: IPAddress) -> typing.Optional[MessageDataSpecifier]
```

Reply:
>With this port to multicast, there's no need for the property local_ip_address which can be replaced by domain_id.

Yes.

>Can the same be said for the functions related to message?

Yes.